src/lib.rs

```diff
@@ -1,350 +1,434 @@
 use std::collections::HashMap;
+use std::collections::HashSet;
+use std::hash::Hash;
 
-/// Compute the gini impurity of a dataset. 
-/// 
+/// Compute the gini impurity of a dataset.
+///
 /// Returns a float, 0 representing a perfectly pure dataset. Normal distribution: ~0.33
-/// 
+///
 /// By default, any empty dataset will return a gini of 1.0. This may be unexpected behaviour.
 /// ```
 /// use parsnip::gini;
-/// assert_eq!(gini(&vec![0, 0, 0, 1]), 0.375);
+/// assert_eq!(gini(&vec![0_usize, 0, 0, 1]), 0.375);
 /// ```
-pub fn gini(data: &[u64]) -> f32 {
-    if data.len() == 0 {
+pub fn gini<T>(data: &[T]) -> f32
+where
+    T: Eq,
+    T: Hash,
+{
+    if data.is_empty() {
         return 1.0;
-    } 
+    }
     fn p_squared(count: usize, len: f32) -> f32 {
         let p = count as f32 / len;
         p * p
     }
     let len = data.len() as f32;
     let mut count = HashMap::new();
-    for &value in data {
+    for value in data.iter() {
```

Comment:
Explicit `iter()` shouldn't be necessary here. Can you try `for value in data {` or `for value in &data {`?

Reply:
`for value in data {` passes!